### PR TITLE
Simplify itemType class and get rid of magic strings for types

### LINKED ISSUE #521: Annotate() fn: Refactor magic type names with ItemType.toString()

### PR DESCRIPTION
**Pull request by CanBerker**
Fixes #521 
- ItemType enumhas been made the single source of truth for itemtypes and their string representations (for both construction and serialization)
- The class has been converted to an enum to achieve referential equality that stays consistent even across deep copying of SequenceTypes. 
- ItemTypes enum is now redundant
- ItemTypes defined in Functions.java are cleaned up as they are redundant.
- Magic values for itemtype-strings in RumbleML are cleaned up
- RumbleML schema itemtype keywords are updated to match the grammar which is 'lowerCamelCase'
- OrderByClauseRuntimeIterator and UDFs are also cleaned up from the magic type strings.

### CODE REVIEW

File: src/main/java/org/rumbledb/items/FunctionItem.java [hunk 1]
```diff
@@ -237,4 +238,19 @@ public void read(Kryo kryo, Input input) {
...
+            oos.writeObject(this);
+            oos.flush();
+            byte[] data = bos.toByteArray();
+            ByteArrayInputStream bis = new ByteArrayInputStream(data);
+            ObjectInputStream ois = new ObjectInputStream(bis);
+            return (FunctionItem) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) {
+            throw new OurBadException("Error while deep copying the function body runtimeIterator");
+        }
+    }
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/items/FunctionItem.java]: Exracted to a method, cosmetic

File: src/main/java/org/rumbledb/runtime/functions/base/Functions.java [hunk 2]
```diff
@@ -295,98 +295,66 @@
...
-        itemTypes.put("date", ItemType.dateItem);
-        itemTypes.put("time", ItemType.timeItem);
-
-        itemTypes.put("hexBinary", ItemType.hexBinaryItem);
-        itemTypes.put("base64Binary", ItemType.base64BinaryItem);
-
-        itemTypes.put("null", ItemType.nullItem);
-
-    }
-
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/runtime/functions/base/Functions.java]: These were redundant as we have static final instances of ItemTypes.

**Comment by ghislainfourny** [comment]
@CanBerker thanks for the proposal -- let us discuss it at the next meeting.

While it is true that, in the current state, an enum does the trick, this will not be true in the future (user-defined types, function types). I am thus afraid that this refactoring may not be compatible with the future evolution of Rumble.

**Comment by CanBerker** [comment]
Sounds good. Looking forward to it

**Comment by ghislainfourny** [comment]
Thanks! By the way, some of the suggested changes in this PR that are orthogonal to making the class an enum and can probably be taken over, so this work was useful.

**Comment by ghislainfourny** [comment]
@CanBerker would it be possible to just revert, in this PR, the conversion of the class to an enum, i.e., convert back the enum values to static constants for forward-compatibility? I think that this should still preserve many of your other (useful) changes. Then I will take a look again. Thanks!

**Comment by CanBerker** [comment]
I was previously pushing to enforce types as singletons that do not get recreated at any part of the program. Enums were very helpful in ensuring this across serialization/deserialization which deep copying also uses. However, with the new implementation, even though i managed to get Java serialization to preserve the singleton nature, I believe kryo remained problematic.

Following these efforts, I have simply implemented an equals method that performs a name check  and dropped the effort into preserving the singleton pattern. As you have mentioned further changes with user defined types, this current implementation should hopefully be sufficient.

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 3]
```diff
@@ -819,7 +819,7 @@ public SequenceType processSequenceType(JsoniqParser.SequenceTypeContext ctx) {
         if (ctx.item == null) {
             return SequenceType.emptySequence;
         }
-        ItemType itemType = new ItemType(ctx.item.getText());
+        ItemType itemType = ItemType.getItemTypeByName(ctx.item.getText());
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: This looks good.

File: src/main/java/org/rumbledb/items/FunctionItem.java [hunk 4]
```diff
@@ -237,4 +238,19 @@ public void read(Kryo kryo, Input input) {
     public ItemType getDynamicType() {
         return ItemType.functionItem;
     }
+
+    public FunctionItem deepCopy() {
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/items/FunctionItem.java]: This looks good too.

File: src/main/java/org/rumbledb/items/parsing/ItemParser.java [hunk 5]
```diff
@@ -292,51 +293,58 @@ private static void addValue(
...
+        if (itemTypeName.equals(ItemType.dateTimeItem.getName())) {
+            return DataTypes.TimestampType;
+        }
+        if (itemTypeName.equals(ItemType.hexBinaryItem.getName())) {
+            return DataTypes.BinaryType;
+        }
+        if (itemTypeName.equals("object")) {
+            return vectorType;
+        }
+        throw new IllegalArgumentException("Unexpected item type found: '" + itemTypeName + "'.");
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/items/parsing/ItemParser.java]: This looks good

File: src/main/java/org/rumbledb/items/parsing/ItemParser.java [hunk 6]
```diff
@@ -292,51 +293,58 @@ private static void addValue(
...
     }
 
     public static String getItemTypeNameFromDataFrameDataType(DataType dataType) {
         if (DataTypes.BooleanType.equals(dataType)) {
-            return "boolean";
+            return ItemType.booleanItem.getName();
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/items/parsing/ItemParser.java]: Also looks good, and elses can all be removed.

File: src/main/java/org/rumbledb/runtime/flwor/clauses/OrderByClauseSparkIterator.java [hunk 7]
```diff
@@ -40,6 +40,7 @@
 import org.rumbledb.runtime.flwor.expression.OrderByClauseAnnotatedChildIterator;
 import org.rumbledb.runtime.flwor.udfs.OrderClauseCreateColumnsUDF;
 import org.rumbledb.runtime.flwor.udfs.OrderClauseDetermineTypeUDF;
+import org.rumbledb.types.ItemType;
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/flwor/clauses/OrderByClauseSparkIterator.java]: This file also looks good.

File: src/main/java/org/rumbledb/runtime/flwor/udfs/OrderClauseDetermineTypeUDF.java [hunk 8]
```diff
@@ -165,7 +141,7 @@ public OrderClauseDetermineTypeUDF(
                         expressionWithIterator.getIterator().getMetadata()
                 );
             } else {
-                throw new OurBadException("Unexpected type found.");
+                this.result.add(this.nextItem.getDynamicType().getName());
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/flwor/udfs/OrderClauseDetermineTypeUDF.java]: Very nice!

File: src/main/java/org/rumbledb/runtime/functions/FunctionItemCallIterator.java [hunk 9]
```diff
@@ -120,7 +120,7 @@ private void wrapArgumentIteratorsWithTypeCheckingIterators() {
                     TypePromotionIterator typePromotionIterator = new TypePromotionIterator(
                             this.functionArguments.get(i),
                             this.functionItem.getSignature().getParameterTypes().get(i),
-                            "Invalid argument for " + this.functionItem.getIdentifier().getName() + "function. ",
+                            "Invalid argument for " + this.functionItem.getIdentifier().getName() + " function. ",
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/functions/FunctionItemCallIterator.java]: good catch

File: src/main/java/org/rumbledb/runtime/functions/base/Functions.java [hunk 10]
```diff
@@ -295,98 +289,66 @@
 
     private static List<FunctionIdentifier> currentUnsetUserDefinedFunctionIdentifiers;
 
-    private static final Map<String, ItemType> itemTypes;
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/functions/base/Functions.java]: Very nice

**Comment by CanBerker** [comment]
Thanks!

**Pull request merged**